<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Getting read pileup for multiple regions &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/design-tabs.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="FastQC-like quality-checking of FASTQ files" href="fastqc.html" />
    <link rel="prev" title="Working with BAM-files" href="bam_handling.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../menu.html" class="icon icon-home"> bionumpy
          </a>
              <div class="version">
                0.2.11
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Geting started with BioNumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">A 10-minute introduction to BioNumPy</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting read pileup for multiple regions</a></li>
<li class="toctree-l1"><a class="reference internal" href="fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="numpy_operations.html">Using NumPy functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="similarity_measures.html">Computing the similarity between two BED-files</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Getting read pileup for multiple regions</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/subsetting_bed.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="getting-read-pileup-for-multiple-regions">
<h1>Getting read pileup for multiple regions<a class="headerlink" href="#getting-read-pileup-for-multiple-regions" title="Permalink to this headline"></a></h1>
<p>This is a brief example of how you can you can analyse alignments inside a given set of regions. This example assumes that:</p>
<ul class="simple">
<li><p>You have a set of alignments as a bed-file (can be converted from bam using bedtools BamToBed).</p></li>
<li><p>You have a set of regions in another bed-file</p></li>
<li><p>For simplicity, we here assumes both files are sorted by chromosome. If not, <cite>bnp.arithmetics.sort_intervals</cite> may be used.</p></li>
</ul>
<p>We assume the alignments-file is quite big, so that we don’t want to load all of it into memory at once.
Thus, we read the file by chromsome using <cite>bnp.groupby</cite> and create a pileup for each chromosome (a NumPy-array).
This pileup can then easily be analysed, e.g. by slicing it with the regions we are interested in using <cite>npstructures.raggedslice</cite>.
The result is a RaggedArray, which behaves very much like a NumPy matrix. Full example below:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="kn">import</span> <span class="nn">npstructures</span> <span class="k">as</span> <span class="nn">nps</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">reads</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/alignments.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read_chunks</span><span class="p">()</span>
<span class="n">regions</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/test2.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/small.chrom.sizes&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">chrom_sizes</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">name</span><span class="p">):</span> <span class="n">size</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="o">.</span><span class="n">size</span><span class="p">)}</span>

<span class="n">reads_by_chromosome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
<span class="n">regions_by_chromosome</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">regions</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">chromosome_reads</span><span class="p">),</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">chromosome_regions</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">reads_by_chromosome</span><span class="p">,</span> <span class="n">regions_by_chromosome</span><span class="p">):</span>
    <span class="c1"># Get read pileup, i.e. number of reads per base in this chromosome</span>
    <span class="c1"># (we convert the pileup, which is a  to a regular numpy array with to_array())</span>
    <span class="n">read_pileup</span> <span class="o">=</span> <span class="n">bnp</span><span class="o">.</span><span class="n">arithmetics</span><span class="o">.</span><span class="n">get_pileup</span><span class="p">(</span><span class="n">chromosome_reads</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span><span class="o">.</span><span class="n">to_array</span><span class="p">()</span>

    <span class="c1"># Subset the pileup on the regions</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="n">nps</span><span class="o">.</span><span class="n">ragged_slice</span><span class="p">(</span><span class="n">read_pileup</span><span class="p">,</span> <span class="n">chromosome_regions</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">chromosome_regions</span><span class="o">.</span><span class="n">stop</span><span class="p">)</span>

    <span class="c1"># Subset is now a RaggedArray. Each row contains the pileup for each region</span>
    <span class="c1"># We can easily e.g. get the max pileup for each region using axis=-1</span>
    <span class="n">max_per_region</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">subset</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Regions analysed:&quot;</span><span class="p">,</span> <span class="n">chromosome_regions</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Max pileup within each region:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">max_per_region</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mean pileup within each region:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">subset</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
<p>In our small example data, we only have three regions for chromosome 1 and 2 regions on chromosome 2. The output should be:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Chromosome chr1
Regions analysed: Interval with 3 entries
               chromosome                    start                     stop
                     chr1                        5                       10
                     chr1                       15                       20
                     chr1                      100                      110
Max pileup within each region:
[3 1 0]
Mean pileup within each region:
[2.4 1.  0. ]

Chromosome chr2
Regions analysed: Interval with 2 entries
               chromosome                    start                     stop
                     chr2                        1                       10
                     chr2                       25                       35
Max pileup within each region:
[1 3]
Mean pileup within each region:
[0.55555556 3.        ]
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="bam_handling.html" class="btn btn-neutral float-left" title="Working with BAM-files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="fastqc.html" class="btn btn-neutral float-right" title="FastQC-like quality-checking of FASTQ files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>