<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Encodings &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/design-tabs.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Supported file formats" href="supported_file_formats.html" />
    <link rel="prev" title="Working with big data" href="working_with_big_data.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../menu.html" class="icon icon-home"> bionumpy
          </a>
              <div class="version">
                0.2.11
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Geting started with BioNumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">A 10-minute introduction to BioNumPy</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup for multiple regions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/numpy_operations.html">Using NumPy functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Encodings</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#encoding-data-using-a-specific-encoding">Encoding data using a specific encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="#supported-encodings">Supported encodings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#when-having-already-encoded-data-advanced-usage">When having already encoded data (advanced usage)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/genome_arithmetics.html">Genome arithmetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Encodings</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/encoding.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="encodings">
<h1>Encodings<a class="headerlink" href="#encodings" title="Permalink to this headline"></a></h1>
<p>A central concept in BioNumPy is the <em>encoding</em> of data, such as DNA sequence, base qualities, kmers, etc, to memory-efficient data types that are used internally by BioNumPy.</p>
<p>For instance, when asking BioNumPy to store the DNA-sequence <cite>s = bnp.as_encoded_array(“ACGT”, bnp.DNAEncoding</cite>, BioNumPy does not store the letters A, C, T and G, but instead uses an efficient numeric representation of them. However, the user does not need to know how this works internally, and will only need to think about the stored sequence as letters. This is why things like <cite>sequence == “A”</cite> works, even though sequence internally is a numeric array.</p>
<section id="encoding-data-using-a-specific-encoding">
<h2>Encoding data using a specific encoding<a class="headerlink" href="#encoding-data-using-a-specific-encoding" title="Permalink to this headline"></a></h2>
<p>When reading data with <cite>bnp.open</cite>, BioNumPy automatically encodes your data with a suitable encoding (determined by the file format). If you for some reason have data from other sources, e.g. as strings or list of strings, you may use <cite>bnp.as_encoded_array</cite> to encode your data. In this case, you should make sure you use an encoding suitable for your data (see supported encodings below).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">bionumpy</span> <span class="k">as</span> <span class="nn">bnp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ACCT&quot;</span><span class="p">,</span> <span class="s2">&quot;AcaATA&quot;</span><span class="p">,</span> <span class="s2">&quot;ca&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bnp</span><span class="o">.</span><span class="n">as_encoded_array</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="go">encoded_ragged_array([&#39;ACCT&#39;,</span>
<span class="go">                      &#39;ACAATA&#39;,</span>
<span class="go">                      &#39;CA&#39;], AlphabetEncoding(&#39;ACGT&#39;))</span>
</pre></div>
</div>
</section>
<section id="supported-encodings">
<h2>Supported encodings<a class="headerlink" href="#supported-encodings" title="Permalink to this headline"></a></h2>
<p>These are the most common encodings that can be used:</p>
<ul class="simple">
<li><p><cite>bnp.BaseEncoding</cite>: Can encode any string</p></li>
<li><p><cite>bnp.DNAEncoding</cite>: Supports A, C, T and G (not N)</p></li>
<li><p><cite>bnp.encodings.alphabet_encoding.ACTGnEncoding</cite>: Supports N, A, C, T, G</p></li>
<li><p><cite>bnp.encodings.alphabet_encoding.ACUGEncoding</cite>: Supports A, C, U, G</p></li>
<li><p><cite>bnp.encodings.alphabet_encoding.RNAEncoding</cite>: Supports A, C, U, G</p></li>
<li><p><cite>bnp.encodings.alphabet_encoding.AminoAcidEncoding</cite>: Supports all valid AminoAcids</p></li>
</ul>
</section>
<section id="when-having-already-encoded-data-advanced-usage">
<h2>When having already encoded data (advanced usage)<a class="headerlink" href="#when-having-already-encoded-data-advanced-usage" title="Permalink to this headline"></a></h2>
<p>In some cases, you may have already encoded data that you want to use with BioNumPy. In this case, you can wrap your data in BioNumPy’s <cite>EncodedArray</cite> or <cite>EncodedRaggedArray</cite> class, but you will need to be sure that your data is encoded correctly as BioNumPy does not verify this.</p>
<p>For example, if you have already encoded DNA-sequences so that A is 0, C is 1, G is 2 and T is 3, your data is compatible with <cite>bnp.DNAEncoding</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">already_encoded</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bnp</span><span class="o">.</span><span class="n">EncodedArray</span><span class="p">(</span><span class="n">already_encoded</span><span class="p">,</span> <span class="n">bnp</span><span class="o">.</span><span class="n">DNAEncoding</span><span class="p">)</span>
<span class="go">encoded_array(&#39;ACGT&#39;, AlphabetEncoding(&#39;ACGT&#39;))</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="working_with_big_data.html" class="btn btn-neutral float-left" title="Working with big data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="supported_file_formats.html" class="btn btn-neutral float-right" title="Supported file formats" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>