<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Genome arithmetics &mdash; bionumpy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/design-tabs.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="IO" href="io.html" />
    <link rel="prev" title="BnpDataclass" href="bnpdataclass.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../menu.html" class="icon icon-home"> bionumpy
          </a>
              <div class="version">
                0.2.11
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Geting started with BioNumPy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">A 10-minute introduction to BioNumPy</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastq_filtering.html">Filtering FASTQ reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/bam_handling.html">Working with BAM-files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/subsetting_bed.html">Getting read pileup for multiple regions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/fastqc.html">FastQC-like quality-checking of FASTQ files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/gc_content.html">Computing GC content inside genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/numpy_operations.html">Using NumPy functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/position_weight_matrix.html">Position Weight Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/set_of_sequences.html">Simulating sequence datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/similarity_measures.html">Computing the similarity between two BED-files</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">BioNumPy Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../source/reading_files.html">Reading files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/working_with_big_data.html">Working with big data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/encoding.html">Encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html">Supported file formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/supported_file_formats.html#implementing-a-new-file-format">Implementing a new file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/intervals.html">Intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/multiple_data_sources.html">Working with Multiple Files/Data Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../source/broadcastable_functions.html">Broadcastable Functions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="bnpdataclass.html">BnpDataclass</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Genome arithmetics</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#api-documentation">API documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="io.html">IO</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">Sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="streams.html">Streams</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide.html">Developer guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manuscript/index.html">BioNumPy manuscript</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../menu.html">bionumpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../menu.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Genome arithmetics</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modules/genome_arithmetics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genome-arithmetics">
<span id="arithmetics-api"></span><h1>Genome arithmetics<a class="headerlink" href="#genome-arithmetics" title="Permalink to this headline"></a></h1>
<p>The <cite>bnp.arithmetics</cite> module contains functions for analysing genomic intervals.</p>
<section id="api-documentation">
<h2>API documentation<a class="headerlink" href="#api-documentation" title="Permalink to this headline"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.forbes">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">forbes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_sizes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.ChromosomeSize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="../_modules/bionumpy/arithmetics/similarity_measures.html#forbes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bionumpy.arithmetics.forbes" title="Permalink to this definition"></a></dt>
<dd><p>Computes the Forbes similarity index for two sets of intervals.</p>
<dl class="simple">
<dt>chromosome_sizes<span class="classifier">ChromosomeSize</span></dt><dd><p>A ChromosomeSizes, typically from reading a chromosome.sizes file with bnp.open()</p>
</dd>
<dt>intervals_a<span class="classifier">Interval</span></dt><dd><p>Must be sorted. Can be read using bnp.open on a bed-file.</p>
</dd>
<dt>intervals_a<span class="classifier">Interval</span></dt><dd><p>Must be sorted. Can be read using bnp.open on a bed-file.</p>
</dd>
</dl>
<dl class="simple">
<dt>float</dt><dd><p>The forbes similarity index.</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.arithmetics</span> <span class="kn">import</span> <span class="n">forbes</span><span class="p">,</span> <span class="n">sort_intervals</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.datatypes</span> <span class="kn">import</span> <span class="n">Interval</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Interval</span><span class="o">.</span><span class="n">from_entry_tuples</span><span class="p">([(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Interval</span><span class="o">.</span><span class="n">from_entry_tuples</span><span class="p">([(</span><span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">25</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">40</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_sorted</span> <span class="o">=</span> <span class="n">sort_intervals</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">sort_order</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_sorted</span> <span class="o">=</span> <span class="n">sort_intervals</span><span class="p">(</span><span class="n">b</span><span class="p">,</span> <span class="n">sort_order</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">forbes</span><span class="p">({</span><span class="s2">&quot;chr1&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">},</span> <span class="n">a_sorted</span><span class="p">,</span> <span class="n">b_sorted</span><span class="p">)</span>
<span class="go">5.625</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.jaccard">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">jaccard</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_sizes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.ChromosomeSize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="../_modules/bionumpy/arithmetics/similarity_measures.html#jaccard"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bionumpy.arithmetics.jaccard" title="Permalink to this definition"></a></dt>
<dd><p>Computes the Jaccard similarity index for two sets of intervals.</p>
<dl class="simple">
<dt>chromosome_sizes<span class="classifier">ChromosomeSize</span></dt><dd><p>A ChromosomeSizes, typically from reading a chromosome.sizes file with bnp.open()</p>
</dd>
<dt>intervals_a<span class="classifier">Interval</span></dt><dd><p>Must be sorted. Can be read using bnp.open on a bed-file.</p>
</dd>
<dt>intervals_a<span class="classifier">Interval</span></dt><dd><p>Must be sorted. Can be read using bnp.open on a bed-file.</p>
</dd>
</dl>
<dl class="simple">
<dt>float</dt><dd><p>The forbes similarity index.</p>
</dd>
</dl>
<p>See forbes for examples.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.sort_intervals">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">sort_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals:</span> <span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_key_function:</span> <span class="pre">callable</span> <span class="pre">=</span> <span class="pre">&lt;function</span> <span class="pre">&lt;lambda&gt;&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order:</span> <span class="pre">Optional[List[str]]</span> <span class="pre">=</span> <span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></span><a class="reference internal" href="../_modules/bionumpy/arithmetics/intervals.html#sort_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bionumpy.arithmetics.sort_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Sort intervals on “chromosome”, “start”, “stop”</p>
<dl class="simple">
<dt>intervals<span class="classifier">Interval</span></dt><dd><p>Unsorted intervals</p>
</dd>
</dl>
<dl class="simple">
<dt>Interval</dt><dd><p>Sorted intervals</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.merge_intervals">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">merge_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bionumpy.arithmetics.merge_intervals" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.get_pileup">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">get_pileup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bionumpy.arithmetics.intervals.GenomicRunLengthArray</span></span></span><a class="reference internal" href="../_modules/bionumpy/arithmetics/intervals.html#get_pileup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bionumpy.arithmetics.get_pileup" title="Permalink to this definition"></a></dt>
<dd><p>Get the number of intervals that overlap each position of the chromosome/contig</p>
<p>This uses run length encoded arrays to handle the sparse data that
we get from intervals.</p>
<dl class="simple">
<dt>intervals<span class="classifier">Interval, </span></dt><dd><p>Intervals on the same chromosome/contig</p>
</dd>
<dt>chromosome_size<span class="classifier">int</span></dt><dd><p>size of the chromsome/contig</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.datatypes</span> <span class="kn">import</span> <span class="n">Interval</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bionumpy.arithmetics</span> <span class="kn">import</span> <span class="n">get_boolean_mask</span><span class="p">,</span> <span class="n">get_pileup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">([</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">12</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pileup</span> <span class="o">=</span> <span class="n">get_pileup</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pileup</span><span class="p">)</span>
<span class="go">[0 0 0 1 1 2 2 1 0 0 1 1 0 0 0 0 0 0 0 0]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bionumpy.arithmetics.get_boolean_mask">
<span class="sig-prename descclassname"><span class="pre">bionumpy.arithmetics.</span></span><span class="sig-name descname"><span class="pre">get_boolean_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bionumpy.bnpdataclass.bnpdataclass.Interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/bionumpy/arithmetics/intervals.html#get_boolean_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bionumpy.arithmetics.get_boolean_mask" title="Permalink to this definition"></a></dt>
<dd><p>Get a boolean mask representing where any inteval hits</p>
<p>Uses run length encoded binary arrays to represent the areas
covered by any interval. The mask that is returned supports numpy ufuncs, 
so that you can run logical operations on them s.a. <cite>&amp; | ~</cite> and also 
numpy indexing so you can use it to filter positions and intervals.</p>
<dl class="simple">
<dt>intervals<span class="classifier">Interval</span></dt><dd><p>Intervals on the same chromosome/contig</p>
</dd>
<dt>chromosome_size<span class="classifier">int</span></dt><dd><p>The size of the chromosome/contig</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">([</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">12</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">intervals</span><span class="p">)</span>
<span class="go">Interval with 3 entries</span>
<span class="go">               chromosome                    start                     stop</span>
<span class="go">                     chr1                        3                        8</span>
<span class="go">                     chr1                        5                        7</span>
<span class="go">                     chr1                       10                       12</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mask</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">intervals</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mask</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="go">[0 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 0 0 0 0]</span>
</pre></div>
</div>
<p>Get complement of the mask:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">complement</span> <span class="o">=</span> <span class="o">~</span><span class="n">mask</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">complement</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="go">[1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 1 1 1 1 1]</span>
</pre></div>
</div>
<p>Get the intersections (<cite>&amp;</cite>) and union (<cite>|</cite>) of the mask and another mask</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">other_mask</span> <span class="o">=</span> <span class="n">get_boolean_mask</span><span class="p">(</span><span class="n">Interval</span><span class="p">([</span><span class="s2">&quot;chr1&quot;</span><span class="p">],</span> <span class="p">[</span><span class="mi">9</span><span class="p">],</span> <span class="p">[</span><span class="mi">15</span><span class="p">]),</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intersection</span> <span class="o">=</span> <span class="n">mask</span> <span class="o">&amp;</span> <span class="n">other_mask</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">intersection</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="go">[0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">union</span> <span class="o">=</span> <span class="n">mask</span> <span class="o">|</span> <span class="n">other_mask</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">union</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="go">[0 0 0 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 0]</span>
</pre></div>
</div>
<p>Find wether some positions overlap the mask:
&gt;&gt;&gt; print(other_mask[intervals.start])
[False False  True]</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="bnpdataclass.html" class="btn btn-neutral float-left" title="BnpDataclass" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="io.html" class="btn btn-neutral float-right" title="IO" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Knut Rand.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>